File: probability_simulations/whole_numbers_sum_mod_3.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <random>
using namespace std;

int simulate_slow(int k) {
    vector<int> numbers;
    for (int i = 1; i <= 3 * k + 1; i++) {
        numbers.push_back(i);
    }
    auto rng = default_random_engine {random_device{}()};
    shuffle(begin(numbers), end(numbers), rng);

    int total = 0;
    for (int i = 0; i < 3 * k + 1; i++) {
        total += numbers.at(i);
        if (total % 3 == 0) {
            return 0;
        }
    }
    return 1;
}

int simulate(int k) {
    vector<int> counts = {k + 1, k, k}; // 1, -1, 0
    return 0;
}

int main() {
    int k = 2;
    int iterations = 100000;

    int total = 0;

    for (int i = 0; i < iterations; i++) {
        total += simulate_slow(k);
    }

    cout << "ratio: " << (float) total / iterations << endl;
    cout << "total: " << total << endl;
    return 0;
}

File: probability_simulations/whole_numbers_sum_mod_3.py
import random
import math

def simulate_slow(k):
    numbers = list(range(1, (3 * k + 1) + 1)) 
    random.shuffle(numbers)

    total = 0
    for number in numbers:
        total += number
        if total % 3 == 0:
            return False        
    return True

def simulate(k):
    index_map = {
        1: 0,
        -1: 1,
        0: 2
    }
    counts = [k+1, k, k] # 1, -1, 0
    total = 0
    for _ in range(3 * k + 1):
        number = random.choices([1, -1, 0], weights=[counts[i] / sum(counts) for i in range(3)], k=1)[0]
        counts[index_map[number]] -= 1
        total += number
        if total % 3 == 0:
            return False
    return True


if __name__ == "__main__":
    k = 4
    iterations = 1000000

    good = 0
    for i in range(iterations):
        if simulate_slow(k):
            good += 1
    print("good / iterations", good / iterations)
    print("maybe actual", math.comb(3 * k, k) / (math.factorial(3 * k + 1) / math.factorial(k + 1) / math.factorial(k) / math.factorial(k)))